### PR TITLE
Webpacker

### PR DESCRIPTION
**Pull request by gmq**
Add support for ActiveAdmin in webpacker mode. (#298)

- Uses rollup to generate the javascript used by sprockets
- Updates javascript files to use es6 imports.
- Adds generators to include the appropriate files when installing depending on how ActiveAdmin was installed.

TODO:
- It assumes the npm package is going to be named "activeadmin_addons" How are we going to handle that release?
- ~~Tests~~
- Potassium support will be handled separately.

For testing:
- Clone this branch
- Run `yarn prepublishOnly` inside cloned folder
- In a new Rails installation add activeadmin and add this gem locally
`gem 'activeadmin_addons', path: '../activeadmin_addons'`
- Install activeadmin in webpacker mode
`bundle exec rails g active_admin:install --use_webpacker`
- Install activeadmin addons
`bundle exec rails g activeadmin_addons:install`
(ignore yarn error)
- Install npm package
`yarn add ../activeadmin_addons`

### CODE REVIEW

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:41]: Missing semicolon                                semi<br>Newline required at end of file but not found    eol-last

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:40]: Missing trailing comma                           comma-dangle

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:28]: Strings must use singlequote                     quotes<br>Missing trailing comma                           comma-dangle

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:27]: Strings must use singlequote                     quotes

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:26]: Strings must use singlequote                     quotes

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:24]: Strings must use singlequote                     quotes

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:21]: Missing semicolon                                semi

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:20]: Missing trailing comma                           comma-dangle

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:19]: Missing trailing comma                           comma-dangle

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:12]: Identifier 'indent_level' is not in camel case   camelcase

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:1]: Definition for rule 'vue/max-len' was not found  vue/max-len

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:35]: Final newline missing.

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:31]: Trailing whitespace detected.

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:15]: Prefer to_s over string interpolation.

File: lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/webpacker/webpacker_generator.rb:8]: Prefer to_s over string interpolation.

File: lib/generators/activeadmin_addons/install/install_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/install/install_generator.rb:5]: Trailing whitespace detected.

File: lib/generators/activeadmin_addons/assets/assets_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/assets/assets_generator.rb:53]: Final newline missing.

File: app/assets/stylesheets/activeadmin_addons/imports/webpack/jquery-datepicker.scss
Comment (monkeyci) [inline app/assets/stylesheets/activeadmin_addons/imports/webpack/jquery-datepicker.scss:1]: Expected single quotes   string-quotes

File: app/assets/stylesheets/activeadmin_addons/all.scss
Comment (monkeyci) [inline app/assets/stylesheets/activeadmin_addons/all.scss:19]: Unexpected missing end-of-source newline   no-missing-end-of-source-newline

File: lib/generators/activeadmin_addons/assets/assets_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/assets/assets_generator.rb:23]: Avoid rescuing without specifying an error class.

File: lib/generators/activeadmin_addons/assets/assets_generator.rb
Comment (monkeyci) [inline lib/generators/activeadmin_addons/assets/assets_generator.rb:12]: Avoid rescuing without specifying an error class.

File: app/javascript/activeadmin_addons/all.js
Comment (monkeyci) [inline app/javascript/activeadmin_addons/all.js:14]: Newline required at end of file but not found    eol-last

File: app/javascript/activeadmin_addons/all.js
Comment (monkeyci) [inline app/javascript/activeadmin_addons/all.js:1]: Definition for rule 'vue/max-len' was not found  vue/max-len

File: app/assets/stylesheets/activeadmin_addons/imports/jquery-datepicker.scss
Comment (monkeyci) [inline app/assets/stylesheets/activeadmin_addons/imports/jquery-datepicker.scss:1]: Unexpected missing end-of-source newline   no-missing-end-of-source-newline

File: app/assets/javascripts/activeadmin_addons/all.js
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/all.js:751]: Strings must use singlequote                                           quotes

File: app/assets/javascripts/activeadmin_addons/all.js
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/all.js:745]: Strings must use singlequote                                           quotes<br>Missing trailing comma                                                 comma-dangle

File: rollup.config.js
Comment (monkeyci) [inline rollup.config.js:42]: Newline required at end of file but not found    eol-last

**Comment by filipkis** [comment]
Thanks @gmq for this PR. Any update on this? What's still needed for it to get merged and released?

**Comment by nicolasrsande** [comment]
+1 Any updates on this? Can it be merged? I Am willing to help if anything is still needed.

**Comment by hgeorgilas** [comment]
Is this going to be merged soon? 
Thanks

**Comment by guiferrpereira** [comment]
excellent work here. can you tell when this will be merged?

**Comment by gmq** [comment]
I can't give a specific timeline, it still needs to be reviewed by the team and time is scarce these days. But it should be sooner rather than later!

**Review by rjherrera — APPROVED** [review]

**Comment by kwent** [comment]
Looking forward to this ! Much needed !

**Comment by kwent** [comment]
@gmq @rjherrera Thanks for the amazing work! Looks like it's very close to the finish line. How can we help ? Regards

**Review by ldlsegovia — APPROVED** [review]

**Comment by dkniffin** [comment]
@gmq Great work on this! I'd love to start using it. What are we waiting on to merge? A review from @hidalgoegz ?

**Comment by qblanc** [comment]
Very happy to find this issue, unable to use the gem with a Rails 6.0 app because of this issue. Thanks for the great work you're doing, can't wait to test it !

**Comment by gmq** [comment]
We're ready to merge to master!

We'd appreciate any bug reports or other feedback about this so if you'd like to test it before we make a new release, please add the following to your Gemfile:

```gem 'activeadmin_addons', git: 'https://github.com/platanus/activeadmin_addons'```

**Comment by agrobbin** [comment]
@gmq this is great! Out of curiosity, are there also plans to remove the hard dependency on `sassc` and `sassc-rails` from the [`gemspec`](https://github.com/platanus/activeadmin_addons/blob/db8ee1394f06243785983410406da22c19b0cd8f/activeadmin_addons.gemspec#L24-L25)? Active Admin [just did that](https://github.com/activeadmin/activeadmin/pull/6584) as well.

**Comment by djalmaaraujo** [comment]
@gmq I can't make it work on activeadmin latest version. Just says activeadmin_addons/all cannot be found. yarn was added, etc. Javascript file is imported but SCSS does not work even with sass-rails on the gemfile.

**Comment by jasonperrone** [comment]
I'm concerned about the comment on this change "if ActiveAdmin was installed in that mode". What about the situation where ActiveAdmin was not installed in that mode, but was later changed to use webpacker?

**Comment by matiasgarcia** [comment]
Same as @djalmaaraujo I can't make it work.

Node module is:

Screenshot from 2021-03-08 11-46-36

I tried doing `@import "activeadmin_addons"` instead of `@import "activeadmin_addons/all"` and now I get this error:

```
ERROR in ./app/javascript/stylesheets/active_admin.scss (./node_modules/css-loader/dist/cjs.js??ref--6-1!./node_modules/postcss-loader/src??ref--6-2!./node_modules/sass-loader/dist/cjs.js??ref--6-3!./app/javascript/stylesheets/active_admin.scss)
Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
SassError: Invalid CSS after "i": expected 1 selector or at-rule, was "import 'select2';"
        on line 1 of node_modules/activeadmin_addons/src/all.js
        from line 13 of /home/mgarcia/projects/withdrawals/app/javascript/stylesheets/active_admin.scss
>> import 'select2';
```

Which fails in the first line of `all.scss` that does:

```
@import 'select2';
```

**Comment by gmq** [comment]
> @gmq I can't make it work on activeadmin latest version. Just says activeadmin_addons/all cannot be found. yarn was added, etc. Javascript file is imported but SCSS does not work even with sass-rails on the gemfile.

> Same as @djalmaaraujo I can't make it work.

I just merged a fix by @ParamagicDev for the documentation. The proper way to import the scss is to use `@import 'activeadmin_addons/src/stylesheets/all';`

> I'm concerned about the comment on this change "if ActiveAdmin was installed in that mode". What about the situation where ActiveAdmin was not installed in that mode, but was later changed to use webpacker?

It checks `config.use_webpacker` in the Active Admin initializer so the addons generator should make the appropriate changes as long as it's enabled there.

**Pull request merged**
